client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,352 @@
+	decoder := pools.TagDecoder().Get()
+	decoder.Reset(encodedTags)
+
+	tsID := pools.CheckedBytesWrapper().Get(elem.ID)
+	nsID := pools.CheckedBytesWrapper().Get(elem.NameSpace)
+	seriesIter := pools.SeriesIterator().Get()
+	seriesIter.Reset(pools.ID().BinaryID(tsID), pools.ID().BinaryID(nsID),
+		decoder, accum.startTime, accum.endTime, iters)
+
+	return seriesIter
+}
+
+func (accum *fetchTaggedResultAccumulator) AsEncodingSeriesIterators(
+	limit int, pools fetchTaggedPools,
+) (encoding.SeriesIterators, bool, error) {
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+
+	numElements := 0
+	accum.responses.forEach(func(_ fetchTaggedIDResults) bool {
+		numElements++
+		return numElements < limit
+	})
+
+	result := pools.MutableSeriesIterators().Get(numElements)
+	result.Reset(numElements)
+	count := 0
+	accum.responses.forEach(func(elems fetchTaggedIDResults) bool {
+		seriesIter := accum.sliceResponsesAsSeriesIter(pools, elems)
+		result.SetAt(count, seriesIter)
+		count++
+		return count < limit
+	})
+
+	exhaustive := accum.exhaustive && count < limit
+	return result, exhaustive, nil
+}
+
+func (accum *fetchTaggedResultAccumulator) AsIndexQueryResults(
+	limit int,
+	pools fetchTaggedPools,
+) (index.QueryResults, error) {
+	var (
+		iter  = newFetchTaggedResultsIndexIterator(pools)
+		count = 0
+	)
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+	accum.responses.forEach(func(elems fetchTaggedIDResults) bool {
+		iter.backing.ids = append(iter.backing.ids, elems[0].ID)
+		iter.backing.nses = append(iter.backing.nses, elems[0].NameSpace)
+		iter.backing.tags = append(iter.backing.tags, elems[0].EncodedTags)
+		count++
+		return count < limit
+	})
+
+	return index.QueryResults{
+		Exhaustive: accum.exhaustive && count < limit,
+		Iterator:   iter,
+	}, nil
+}
+
+type fetchTaggedShardConsistencyResults []fetchTaggedShardConsistencyResult
+
+func (res fetchTaggedShardConsistencyResults) initialize(length int) fetchTaggedShardConsistencyResults {
+	if cap(res) < length {
+		res = append(res, make(fetchTaggedShardConsistencyResults, length-cap(res))...)
+	}
+	res = res[:length]
+	// following compiler optimized memcpy impl:
+	// https://github.com/golang/go/wiki/CompilerOptimizations#optimized-memclr
+	for i := range res {
+		res[i] = fetchTaggedShardConsistencyResult{}
+	}
+	return res
+}
+
+type fetchTaggedIDResults []*rpc.FetchTaggedIDResult_
+
+// the returned bool indicates if the iteration should be continued
+type forEachFetchTaggedIDFn func(responsesForSingleID fetchTaggedIDResults) bool
+
+// forEach iterates over the provide results, and calls `fn` on each
+// group of responses with the same ID.
+// NB: assumes the results array being operated upon has been sorted
+func (results fetchTaggedIDResults) forEach(fn forEachFetchTaggedIDFn) {
```

Comment:
do you think there is value in calling this `forEachSorted` or is that just more confusing?

Reply:
I take it back after reading the code. Now I think `forEachGroup` would be better.

Also, would it be too expensive to wrap the results in a struct that contains a `sorted` bool and then have this function panic if its called on an unsorted list?

Reply:
bump. Cool if you disagree just making sure you saw it, I don't have super strong feelings about it

Reply:
`forEachGroup` is more confusing to me, i'll rename to `forEachID` in the next patch.

Reply:
done